tower-log/src/lib.rs

```diff
@@ -0,0 +1,452 @@
+//! transformed into another error type that might provide less
+//! information, such as by `tower-buffer`.
+
+extern crate futures;
+extern crate tower;
+extern crate log;
+
+use futures::{Async, Future, Poll};
+use tower::{Service, NewService};
+
+use std::error::Error;
+use std::fmt;
+
+/// Wrap a `Service` or `NewService` with `LogErrors` middleware.
+///
+/// Unlike using `LogErrors::new`, this macro will configure the returned
+/// middleware to log messages with the module path and file/line location of
+/// the _call site_, as using the `log!` macro in that file would.
+///
+/// For example:
+/// ```rust,ignore
+/// mod my_module {
+///     fn my_function<S>(service: S) -> LogErrors<S>
+///     where S: Service,
+///           S::Error: Error,
+///     {
+///         let timeout = Timeout::new(
+///             service, timer, Duration::from_secs(1)
+///         );
+///         log_errors!(timeout)
+///     }
+/// }
+/// ```
+/// will produce log messages like
+/// ```notrust,ignore
+/// ERROR 2018-03-05T18:36:36Z: my_crate::my_module: Future::poll: operation timed out after Duration { secs: 1, nanos: 0 }
+/// ```
+/// while
+/// ```rust,ignore
+/// mod my_module {
+///     fn my_function<S>(service: S) -> LogErrors<S>
+///     where S: Service,
+///           S::Error: Error,
+///     {
+///         let timeout = Timeout::new(
+///             service, timer, Duration::from_secs(1)
+///         );
+///         LogErrors::new(timeout)
+///     }
+/// }
+/// ```
+/// will produce log messages like
+/// ```notrust,ignore
+/// ERROR 2018-03-05T18:36:36Z: tower_log_errors: Future::poll: operation timed out after Duration { secs: 1, nanos: 0 }
+/// ```
+#[macro_export]
+macro_rules! log_errors {
+    ($inner:expr) => {
+        log_errors!(level: ::log::Level::Error, $inner)
+    };
+    (target: $target:expr, $($rest:tt)+) => {
+        log_errors!($($rest),+).target($target)
+    };
+    (level: $level:expr, $inner:expr) => {
+        $crate::LogErrors::new($inner)
+            .in_module(module_path!())
+            .at_location(file!(), line!())
+            .at_level($level)
+    };
+}
+
+/// Wrap a `Service` or `NewService` with `LogResponses` middleware.
+///
+/// Unlike using `LogResponses::new`, this macro will configure the returned
+/// middleware to log messages with the module path and file/line location of
+/// the _call site_, as using the `log!` macro in that file would.
+#[macro_export]
+macro_rules! log_responses {
+    ($inner:expr) => {
+        log_responses!(level: ::log::Level::Debug, $inner)
+    };
+    (target: $target:expr, $($rest:tt)+) => {
+        log_responses!($($rest),+).target($target)
+    };
+    (level: $level:expr, $inner:expr) => {
+        $crate::LogResponses::new($inner)
+            .in_module(module_path!())
+            .at_location(file!(), line!())
+            .at_level($level)
+    };
+}
+/// Logs error responses.
+#[derive(Clone, Debug)]
+pub struct LogErrors<T> {
```

Comment:
I'm not certain if `LogErrors` and `LogResponses` should be separate instead of making a single `Log` service. I'm thinking having fewer types would be better here (i.e, a single `Log` service).